### PR TITLE
[gRPC/Source] Implement gRPC tensor source

### PR DESCRIPTION
**Pull request by dongju-chae**
This patch implements gRPC tensor source and changes relevant build scripts.

Target gRPC version: `v1.20.1` (on `xenial`, `bionic`, and `tizen`)
Related issue: https://github.com/nnstreamer/nnstreamer/issues/2625

@gichan-jang Please check that this patch for protobuf dependency is okay to your impl (e.g., protobuf converter).

Signed-off-by: Dongju Chae <dongju.chae@samsung.com>

### CODE REVIEW

File: ext/nnstreamer/meson.build [hunk 1]
```diff
@@ -6,6 +6,21 @@ if flatbuf_support_is_available
...
+      output: ['@BASENAME@.grpc.pb.h', '@BASENAME@.grpc.pb.cc'],
+      arguments : [
+        '--proto_path=@CURRENT_SOURCE_DIR@/include',
+        '--plugin=protoc-gen-grpc=' + grpc_cpp_plugin_path.stdout().strip(),
+        '--grpc_out=@BUILD_DIR@',
+        '@INPUT@'
+      ]
+    )
+    grpc_pb_gen_src = grpc_pb_gen.process('./include/nnstreamer.proto')
+  endif
```
Comment (myungjoo) [inline ext/nnstreamer/meson.build:22]: Please properly handle all 4 cases: (protobuf_support, grpc_support) = (0,0), (1,1), (0,1), (1,0), along with proper error/warning messages

Comment (dongju-chae) [reply ext/nnstreamer/meson.build:22]: Sure, got it.

**Comment by myungjoo** [comment]
Reference? https://github.com/technicianted/gsttransformer

**Comment by dongju-chae** [comment]
PR updated: impl. of gRPC/protobuf support are moved to  `ext/nnstreamer/extra/...` because dynamic-loaded `libprotobuf` does not allow register `nnstreamer.proto` twice.

File: ext/nnstreamer/registerer/grpc.c [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation;
+ * version 2.1 of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
```
Comment (myungjoo) [inline ext/nnstreamer/registerer/grpc.c:13]: Use SPDX for new files

File: ext/nnstreamer/tensor_source/tensor_src_grpc.c [hunk 3]
```diff
@@ -0,0 +1,553 @@
...
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation;
+ * version 2.1 of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_source/tensor_src_grpc.c:13]: Use SPDX for new files

File: ext/nnstreamer/registerer/grpc.c (hunk 2 above)
Comment (dongju-chae) [reply ext/nnstreamer/registerer/grpc.c:13]: Got it, thanks.

File: ext/nnstreamer/tensor_source/tensor_src_grpc.c (hunk 3 above)
Comment (dongju-chae) [reply ext/nnstreamer/tensor_source/tensor_src_grpc.c:13]: Got it, thanks.

File: packaging/nnstreamer.spec [hunk 4]
```diff
@@ -184,6 +185,10 @@ BuildRequires:  pkgconfig(ovxlib)
 BuildRequires:  pkgconfig(amlogic-vsi-npu-sdk)
 %endif
 
+%if 0%{?grpc_support}
+BuildRequires:  grpc-devel
```
Comment (myungjoo) [inline packaging/nnstreamer.spec:189]: CHeck if grpc is available in 1. F-Hub/DA and 2. VD.
If not available, add GBM conditions (Refer to LINE 20, 62, 68)

Comment (dongju-chae) [reply packaging/nnstreamer.spec:189]: Oh, I think its version might be not matched. I will disable grpc in both profiles. Thanks.

File: ext/nnstreamer/registerer/grpc.c [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+ */
+
+/**
+ * @file	grpc.c
+ * @date	22 Oct 2020
+ * @brief	Registers nnstreamer extension plugin for gRPC
+ * @see		https://github.com/nnstreamer/nnstreamer
```
Comment (wooksong) [inline ext/nnstreamer/registerer/grpc.c:21]: Please check your indentation.

File: ext/nnstreamer/registerer/grpc.c [hunk 6]
```diff
@@ -0,0 +1,61 @@
+/* SPDX-License-Identifier: LGPL-2.1-only */
+/**
+ * nnstreamer registerer for gRPC plugin
+ * Copyright (C) 2020 Dongju Chae <dongju.chae@samsung.com>
+ *
+ * This library is free software; you can redistribute it and/or
```
Comment (wooksong) [inline ext/nnstreamer/registerer/grpc.c:6]: You may remove these header comments. IMO, the SPDX identifier above is enough.

File: ext/nnstreamer/extra/nnstreamer_protobuf_grpc.cc [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+
+      /* initiate the RPC call */
+      std::unique_ptr< ClientWriter<Tensors> > writer(
+          stub_->SendTensors (&context, &reply));
+
+      GstDataQueueItem *item;
+      GstBuffer *buffer;
+
+      /* until flushing */
+      while (gst_data_queue_pop (queue_, &item)) {
```
Comment (wooksong) [inline ext/nnstreamer/extra/nnstreamer_protobuf_grpc.cc:233]: Q: Is it possible to block forever?

File: ext/nnstreamer/registerer/grpc.c (hunk 5 above)
Comment (dongju-chae) [reply ext/nnstreamer/registerer/grpc.c:21]: Got it

File: ext/nnstreamer/extra/nnstreamer_protobuf_grpc.cc (hunk 7 above)
Comment (dongju-chae) [reply ext/nnstreamer/extra/nnstreamer_protobuf_grpc.cc:233]: Well, it's unblocked before `stop` is called. I'm not sure that it's always guaranteed in gstreamer internals.

File: ext/nnstreamer/registerer/grpc.c (hunk 6 above)
Comment (dongju-chae) [reply ext/nnstreamer/registerer/grpc.c:6]: Got it.

File: ext/nnstreamer/extra/nnstreamer_protobuf_grpc.cc [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+    {
+      gst_tensors_config_init (&config_);
+    }
+
+    /** @brief destructor */
+    ~NNStreamerRPC () {
+      if (server_.get ())
+        server_->Shutdown ();
+
+      if (worker_.joinable ())
```
Comment (wooksong) [inline ext/nnstreamer/extra/nnstreamer_protobuf_grpc.cc:58]: Note: In my case, join () in the destructor () is never passed. (https://github.com/nnstreamer/nnstreamer-ros/pull/56/files#diff-e85c333c6471ef6fc92208770d33712586b9c5974a98113adfad5f1213a40ae1R173) Please check this out.

**Comment by gichan-jang** [comment]
Q. If `nnstreamer_protobuf.so` and `nnstreamer_protobuf_grpc.so`  are loaded at the same time.
Is there any problem if grpc tensor_source  is used with tensor_converter (protobuf->tensor)?

**Comment by dongju-chae** [comment]
@gichan-jang No, I resolved this issue. I checked `tensor_converter` does not have the grpc dependency.

**Review by gichan-jang — APPROVED** [review]
LGTM

**Review by anyj0527 — APPROVED** [review]
LGTM THANKS!
Build checked with grpc-v1.20.1 and protobuf-v3.7.0

**Pull request merged**